### ISSUE TITLE
[design flaw] user config, plugin api and application api are mutually exclusive

### ISSUE THREAD

**Issue by samuelstroschein** [inlang/inlang#1140]
## Problem 

The observation that we need:

1. a "user config" that must be CRUD operable. 
2. A plugin API that exposes stuff like #1116
3. an inlang instance with "app API" #1122

1 <-> 2 can't (?) be identical to enable simple CRUD ops. Hence, a singular `.js` config that combines "user" and "plugin config" is likely wrong. 

Furthermore:

- .js config is not easily typesafe
- .js config leads to design flaws like #1139 because the full power of code can be used (but is never used)

## Proposal 

Split inlang core APIs into three: 

1. A user-defined config. 
2. A plugin API that defines callbacks. 
3. An application API (an inlang app, see https://github.com/inlang/inlang/issues/1122#issuecomment-1641898910)

```
- core
  - plugin
    - api.ts
    - createPlugin.ts
  - config
    - schema.json
    - parseConfig.ts
  - app
    - api.ts
    - createApp.ts
```

#### 1. A user defined config (JSON)

- enable CRUD changes via apps
- simplify the config (no more env, fs, etc)
- achieve typesafety via https://json-schema.org/ (native support in VSCode)
- increase interop (other apps can simply read the json config)

Rough draft of user json config

This: 

```ts
/**
 * @type { import("@inlang/core/config").DefineConfig }
 */
export async function defineConfig(env) {
	const { default: jsonPlugin } = await env.$import(
		"https://cdn.jsdelivr.net/npm/@inlang/plugin-json@3/dist/index.js",
	)

	const { default: standardLintRules } = await env.$import(
		"https://cdn.jsdelivr.net/npm/@inlang/plugin-standard-lint-rules@3/dist/index.js",
	)

	return {
		referenceLanguage: "en",
		plugins: [
			jsonPlugin({ pathPattern: "source-code/website/lang/{language}.json" }),
			standardLintRules(),
		],
	}
}
```

Could turn into this:

```json

{
	"referenceLanguageTag": "en",
	"plugins": {
		"inlang.standardLintRules": {
			"module": "https://cdn.jsdelivr.net/npm/@inlang/plugin-standard-lint-rules@3/dist/index.js"
		},
		"inlang.plugin-json": {
			"module": "https://cdn.jsdelivr.net/npm/@inlang/plugin-json@3/dist/index.js",
			"options": {
				"pathPattern": "source-code/website/lang/{language}.json"
			}
		}
	}
}

```

#### 2. Have a dedicated plugin API (mostly what we have so far in the inlang config js). 

- already have that a plugin can define "config props" 
- but callbacks like `loadMessages` and `saveMessages` shouldn't be in user config then 

https://github.com/inlang/inlang/blob/be7e195a955994f463e59a9a39d9b2172da6549b/source-code/core/src/plugin/types.ts#L25-L31

#### 3. Inlang instance API 

From https://github.com/inlang/inlang/issues/1122#issuecomment-1641898910

> New proposal https://www.loom.com/share/13d70a3063ea4959b058fb19c5953556
> 
> * `setupConfig` should return an "inlang app", not a config -> `createInlangApp`
> * the inlang instance has getters that can be subscribed to (solving the reactive problem)
> * inlang instance deals with caching, calls `loadMessages()` and `saveMessages() internally as required.

## Additional information 

- with #1122 we are separating the need for apps to directly interact with the config. hence, abstract the whole env, fs stuff etc.

- the JS config was easy to get started because easily hackable with exposing of `fs` and manual `import`. but the reality is that plugins are now the main driver. exposing `$fs` and `$import` in the config is unnecessary. 

- This is a massive change. But, might be worth pulling it off now while we can update configs manually via pull requests

**Comment by felixhaeberle** [comment]
When I look at the bigger picture of this change, the architecture is much more comprehensive & easier. 

But: We have to flesh out the details of those early thoughts because some things are yet not clearly defined and therefore can't be commented on.

Personally, I always wondered:
- inlang config is code (js)
- languages not declarative
- inlang has no global API like inlang.messages (only top level functions readResources/ writeResources) 
- setupConfig process in apps has no importance, complicates it, could be abstracted 

Those irritation's -> they are all fixed with this proposal.

**Comment by ivanhofer** [comment]
> * exposing `$fs` and `$import` in the config is unnecessary

But we still need them in the plugins, right?

**Comment by samuelstroschein** [comment]
will be addressed in #1149
